node/chainSimulator/chainSimulator_test.go

```diff
@@ -157,6 +161,157 @@ func TestChainSimulator_SetState(t *testing.T) {
+func TestChainSimulator_AddValidatorKey(t *testing.T) {
+	startTime := time.Now().Unix()
+	roundDurationInMillis := uint64(6000)
+	roundsPerEpoch := core.OptionalUint64{
+		HasValue: true,
+		Value:    20,
+	}
+	chainSimulator, err := NewChainSimulator(ArgsChainSimulator{
+		BypassTxSignatureCheck: false,
+		TempDir:                t.TempDir(),
+		PathToInitialConfig:    defaultPathToInitialConfig,
+		NumOfShards:            3,
+		GenesisTimestamp:       startTime,
+		RoundDurationInMillis:  roundDurationInMillis,
+		RoundsPerEpoch:         roundsPerEpoch,
+		ApiInterface:           api.NewNoApiInterface(),
+		MinNodesPerShard:       3,
+		MetaChainMinNodes:      3,
+	})
+	require.Nil(t, err)
+	require.NotNil(t, chainSimulator)
+
+	err = chainSimulator.GenerateBlocks(30)
+	require.Nil(t, err)
+
+	// add validator key
+	validatorKeys := &dtos.ValidatorsKeys{
+		PrivateKeysBase64: []string{"NjRhYjk3NmJjYWVjZTBjNWQ4YmJhNGU1NjZkY2VmYWFiYjcxNDI1Y2JiZDcwYzc1ODA2MGUxNTE5MGM2ZjE1Zg=="},
+	}
+	err = chainSimulator.AddValidatorKeys(validatorKeys)
+	require.Nil(t, err)
+
+	newValidatorOwner := "erd1l6xt0rqlyzw56a3k8xwwshq2dcjwy3q9cppucvqsmdyw8r98dz3sae0kxl"
+	newValidatorOwnerBytes, _ := chainSimulator.nodes[1].GetCoreComponents().AddressPubKeyConverter().Decode(newValidatorOwner)
+	rcv := "erd1qqqqqqqqqqqqqqqpqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqplllst77y4l"
+	rcvAddrBytes, _ := chainSimulator.nodes[1].GetCoreComponents().AddressPubKeyConverter().Decode(rcv)
+
+	// set balance for sender
+	err = chainSimulator.SetStateMultiple([]*dtos.AddressState{
+		{
+			Address: "erd1l6xt0rqlyzw56a3k8xwwshq2dcjwy3q9cppucvqsmdyw8r98dz3sae0kxl",
+			Balance: "10000000000000000000000",
+		},
+	})
+	require.Nil(t, err)
+
+	blsKey := "9b7de1b2d2c90b7bea8f6855075c77d6c63b5dada29abb9b87c52cfae9d4112fcac13279e1a07d94672a5e62a83e3716555513014324d5c6bb4261b465f1b8549a7a338bc3ae8edc1e940958f9c2e296bd3c118a4466dec99dda0ceee3eb6a8c"
+	blsKeyBytes, _ := hex.DecodeString(blsKey)
+	privateKey := chainSimulator.nodes[0].GetCryptoComponents().KeysHandler().GetHandledPrivateKey(blsKeyBytes)
+	signedMessage, _ := chainSimulator.nodes[0].GetCryptoComponents().BlockSigner().Sign(privateKey, newValidatorOwnerBytes)
+
+	// stake validator
+	stakeValue, _ := big.NewInt(0).SetString("2500000000000000000000", 10)
+	tx := &transaction.Transaction{
+		Nonce:     0,
+		Value:     stakeValue,
+		SndAddr:   newValidatorOwnerBytes,
+		RcvAddr:   rcvAddrBytes,
+		Data:      []byte(fmt.Sprintf("stake@01@%s@%s", blsKey, hex.EncodeToString(signedMessage))),
+		GasLimit:  50_000_000,
+		GasPrice:  1000000000,
+		Signature: []byte("dummy"),
+		ChainID:   []byte(configs.ChainID),
+		Version:   1,
+	}
+
+	err = chainSimulator.nodes[1].GetFacadeHandler().ValidateTransaction(tx)
+	require.Nil(t, err)
+
+	_, err = chainSimulator.nodes[1].GetFacadeHandler().SendBulkTransactions([]*transaction.Transaction{tx})
+	require.Nil(t, err)
+
+	time.Sleep(100 * time.Millisecond)
+
+	err = chainSimulator.GenerateBlocks(5)
```

Comment:
can we add comments containing steps? This test is long and hard to understand.

Reply:
done